File: 5.最长回文子串.cpp
// dp
class Solution {
public:
    string longestPalindrome(string s) {
        int len = s.size();
        vector<vector<int>> isp(len, vector<int>(len, 0));
        for(int right = 0; right < len; ++right){
            for(int left = 0; left <= right; ++left)
                isp[left][right] = s[left] == s[right] && (right - left <= 2 || isp[left + 1][right - 1]);
        }
        int start = 0, end = 0;
        int maxLen = 0;
        for(int i = 0; i < len; ++i){
            for(int j = 0; j < len; ++j){
                if(isp[i][j]){
                    if(maxLen < j - i + 1){
                        maxLen = j - i + 1;
                        start = i;
                        end = j;
                    }
                }
            }
        }
        return s.substr(start, end - start + 1);
    }
};

File: LeetCode 热题 HOT 100/5.最长回文子串.js
/**
 * @param {string} s
 * @return {string}
 */
 var longestPalindrome = function (s) {
    let n = s.length;
    // dp[i][j] 表示子串 s[i ~ j]是否是回文串
    let dp = new Array(n).fill(0).map(() => { return new Array(n).fill(false) });
    let maxLen = 0, start = 0, end = 0;
    for (let i = n - 1; i >= 0; --i) {
        for (let j = i; j < n; ++j) {
            dp[i][j] = (i + 1 <= j - 1 ? dp[i + 1][j - 1] : true) & s.charAt(i) === s.charAt(j);
            if (dp[i][j] && maxLen < (j - i + 1)) {
                maxLen = j - i + 1;
                start = i;
                end = j;
            }
        }
    }
    return s.slice(start, end + 1);
};

// 一维化
var longestPalindrome = function (s) {
    let n = s.length;
    let maxLen = 0, start = 0, end = 0;
    let dp = new Array(n).fill(false);
    dp[n - 1] = true;
    for (let i = n - 2; i >= 0; --i) {
        dp[i] = true;
        for (let j = n - 1; j > i; --j) {
            dp[j] = (i + 1 <= j - 1 ? dp[j - 1] : true) & s.charAt(i) === s.charAt(j);
            if (dp[j] && maxLen < (j - i + 1)) {
                maxLen = j - i + 1;
                start = i;
                end = j;
            }
        }
    }
    return s.slice(start, end + 1);
};